### DISCUSSION TITLE
How to use tasmota.remove_driver correctly?

### DISCUSSION THREAD

**Discussion by Slyke** [arendst/Tasmota#15131] [Q&A] [answered]
Hello, just figuring out how to use Berry Script with LVGL and figuring out classes.

I've made my own class that registers itself with `tasmota.add_driver(myClass)`, but when reuploading my code changes and running `load "autoexec"` again, I can see my old script is still being executed as well (both on screen event hooks and prints `every_second()`).

How can I tell my old code to unregister itself and halt execution with `tasmota.remove_driver` when a new instance is added, and/or manually running a command without restarting the device?

**Comment by sfromis** [comment, marked as answer]
You could have a `reload` command in the driver, telling it to remove the old instance before creating/adding a new. This presumes that the driver was already instantiated when you want to replace it.

**Reply by Slyke** [reply]
Something like:

```
myClass = MyClass()

def reload()
  tasmota.remove_driver(myClass)
  tasmota.remove_cmd('reload')
end

tasmota.add_driver(myClass)
tasmota.add_cmd('reload', reload)
```

Would work?

How would I get the "process" to exit? Technically it wouldn't be doing much any more, but it would be taking up a little bit of memory.

Also, would running `reload` trigger the function on all instances listening to it, or just the last registered one?

Edit: Seems like it only kills the last registered one.

**Reply by sfromis** [reply]
Yeah, it does not care about what's registered....

What do you mean by getting the "process" to exit?

Anyway, what I meant was to have the reload command as part of the class, not needing to rely on any global scope variables. Then it could do a `tasmota.remove_driver(self)`. Getting this to work needs a closure to capture `self` for the remove, instead of directly using a class function as argument to `tasmota.add_cmd()`

Something like:
```
tasmota.add_cmd('reload', 
  def()
    tasmota.remove_driver(self)
    tasmota.remove_cmd('reload')
  end )
```

Of course, you could also add with code to create a new instance, where the `init()` function could do the adds.

Depending on what's convenient to you, you may not need a command-line command, if the instance is housed in `myClass` you could also execute a `myClass.reload()` by other means.

**Reply by Slyke** [reply]
Hmmm, still having problems with this. I stuck it right in the init() for the class

autoexec.be:
```
import DisplayBoot
displayBoot = DisplayBoot()
```

DisplayBoot.be:
```
class DisplayBoot
  def init()
    tasmota.add_cmd('reload',
      def()
        tasmota.remove_driver(self)
        tasmota.remove_cmd('reload')
      end
    )
    tasmota.add_driver(self)
  end
end
```

But it doesn't seem to register the `reload` command in the berry console:
```
reload()
BRY: Exception> 'syntax_error' - input:1: 'reload' undeclared (first use in this function)
```
Even though I can see that it is registering the driver correctly.

**Reply by s-hadinger** [reply]
What you did adds a Tasmota command that works in the Tasmota console. If it's just to use in Berry, just create a function and call it.

**Reply by s-hadinger** [reply]
And one could argue that `reload` is too generic as a command name.

**Reply by sfromis** [reply]
> reload()

What you do there is not a command but a function call, and you cannot issue Tasmota commands in the Berry console.

And if you want a "reload" command, it really should reload, not just unload the old one. Naming is an issue too, what I posted earlier was not a name suggestion, but a generic function suggestion.
